pkg/services/cleanup/cleanup.go

```diff
@@ -272,16 +281,90 @@ func (srv *CleanUpService) expireOldVerifications(ctx context.Context) {
 
 func (srv *CleanUpService) deleteStaleShortURLs(ctx context.Context) {
 	logger := srv.log.FromContext(ctx)
-	cmd := shorturls.DeleteShortUrlCommand{
-		OlderThan: time.Now().Add(-time.Duration(srv.Cfg.ShortLinkExpiration*24) * time.Hour),
-	}
-	if err := srv.ShortURLService.DeleteStaleShortURLs(ctx, &cmd); err != nil {
-		logger.Error("Problem deleting stale short urls", "error", err.Error())
+	if srv.Features.IsEnabledGlobally(featuremgmt.FlagKubernetesShortURLs) {
+		srv.deleteStaleKubernetesShortURLs(ctx)
 	} else {
-		logger.Debug("Deleted short urls", "rows affected", cmd.NumDeleted)
+		cmd := shorturls.DeleteShortUrlCommand{
+			OlderThan: time.Now().Add(-time.Duration(srv.Cfg.ShortLinkExpiration*24) * time.Hour),
+		}
+		if err := srv.ShortURLService.DeleteStaleShortURLs(ctx, &cmd); err != nil {
+			logger.Error("Problem deleting stale short urls", "error", err.Error())
+		} else {
+			logger.Debug("Deleted short urls", "rows affected", cmd.NumDeleted)
+		}
 	}
 }
 
+func (srv *CleanUpService) deleteStaleKubernetesShortURLs(ctx context.Context) {
+	logger := srv.log.FromContext(ctx)
+	logger.Debug("Starting deleting expired Kubernetes shortURLs")
+
+	// Create the dynamic client for Kubernetes API
+	restConfig, err := srv.clientConfigProvider.GetRestConfig(ctx)
+	if err != nil {
+		logger.Error("Failed to get REST config for Kubernetes client", "error", err.Error())
+		return
+	}
+
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		logger.Error("Failed to create Kubernetes client", "error", err.Error())
+		return
+	}
+
+	// Set up the GroupVersionResource for shortURLs
+	gvr := schema.GroupVersionResource{
+		Group:    v1alpha1.ShortURLKind().Group(),
+		Version:  v1alpha1.ShortURLKind().Version(),
+		Resource: v1alpha1.ShortURLKind().Plural(),
+	}
+
+	// Calculate the expiration time
+	expirationTime := time.Now().Add(-time.Duration(srv.Cfg.ShortLinkExpiration*24) * time.Hour)
+	expirationTimestamp := expirationTime.Unix()
+	deletedCount := 0
+
+	// List all shortURLs across all namespaces
+	// TODO: how to target only this instance namespace in a background process?
+	shortURLs, err := client.Resource(gvr).Namespace("default").List(ctx, v1.ListOptions{Limit: 100})
```

Comment:
I need to get the instance namespaces list here. In the cloud, it will be one per instance, but there could be more in on-prem.

Reply:
You can do that by using the org service and searching that:  https://github.com/grafana/grafana/blob/main/pkg/services/dashboards/service/dashboard_service.go#L163